### PR TITLE
Fix suggestion to borrow in struct

### PR DESCRIPTION
**Pull request by kper**
The corresponding issue is #71136.
The compiler suggests that borrowing `the_foos` might solve the problem. This is obviously incorrect.
```
struct Foo(u8);

#[derive(Clone)]
struct FooHolster {
    the_foos: Vec<Foo>,
}
```

I propose as fix to check if there is any colon in the span. However, there might a case where `my_method(B { a: 1, b : foo })` would be appropriate to show a suggestion for `&B ...`.  To fix that too, we can simply check if there is a bracket in the span. This is only possible because both spans are different. 
Issue's span: `the_foos: Vec<Foo>`
other's span: `B { a : 1, b : foo }`

### CODE REVIEW

File: src/librustc_trait_selection/traits/error_reporting/suggestions.rs [hunk 1]
```diff
@@ -591,12 +617,23 @@ impl<'a, 'tcx> InferCtxtExt<'tcx> for InferCtxt<'a, 'tcx> {
...
                     );
-                    err.span_suggestion(
-                        span,
-                        "consider borrowing here",
-                        format!("&{}", snippet),
-                        Applicability::MaybeIncorrect,
-                    );
+
+                    // This if is to prevent a special edge-case
+                    if !has_colon || has_double_colon || has_bracket {
```
Comment (matthewjasper) [inline src/librustc_trait_selection/traits/error_reporting/suggestions.rs:622]: I would prefer this to check for `!span.from_expansion()` than using snippets.

Comment (kper) [reply src/librustc_trait_selection/traits/error_reporting/suggestions.rs]: uhhh thanks. That's much better :)

**Pull request merged**